Add support for easy serialization of container/ranges via StringBuilder

weinig opened the pull request:
#### 7fd3deadf8b1a134082e4420ac140a3227d68d34
<pre>
Add support for easy serialization of container/ranges via StringBuilder
<a href="https://bugs.webkit.org/show_bug.cgi?id=275062">https://bugs.webkit.org/show_bug.cgi?id=275062</a>

Reviewed by Darin Adler.

Introduced a new `StringTypeAdapter`, `Interleave`, which takes
a container/range, a joining string, and an optional mapping function.
`Interleaves` can currently only be used with StringBuilder, but can
be extended to makeString() as well with a bit of header factoring.

`Interleave` replaces common hand written loops when trying to
serialize container/ranges. For example, the following code:

    StringBuilder builder;

    ...

    builder.append(&apos;(&apos;);
    bool wroteFirstItem = false;
    for (auto&amp; item : container) {
        if (wroteFirstItem)
            builder.append(&quot;, &quot;_s);
        wroteFirstItem = true;
        builder.append(item-&gt;value);
    }
    builder.append(&apos;)&apos;);

can be replaced with:

    StringBuilder builder;

    ...

    builder.append(&apos;(&apos;, interleave(container, [](auto&amp; item) { return item-&gt;value; }, &quot;, &quot;_s), &apos;)&apos;);

`Interleave` works via a new kind of `StringTypeAdapter` that doesn&apos;t
know the length/8-bit-ness of its value before hand, causing the
builder to go down a new &quot;slow&quot; path. This &quot;slow&quot; path is just as
efficient as the old hand written loop, but a bit more ergonomic.

In addition to introducing `Interleave`, it has been adopted in places
in WebCore where it can be deployed (that I found), but has not been
adopted in places where adopting it would have turned the builder usage
into a one-liner. Those will be become makeString() calls when added.

* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::appendFromAdapters):
(WTF::StringBuilder::appendFromAdapterSlow):
(WTF::StringBuilder::appendFromAdaptersSlow):
    - Detect use of a &quot;slow&quot; `StringTypeAdapter` and use the
      alternative path, iteratively appending each adapter.

      The current strategy checks if any of the remaining
      `StringTypeAdapters` are &quot;slow&quot; and goes down the
      alternative path, peeling off the first adapter and
      appending it, and then recursing back to the beginning
      with the remainder of the adapters. This means that
      if an append() contains:

          append(&apos;(&apos;, 10, interleave(range, &apos; &apos;), hex(10), &apos;)&apos;); [ aka &quot;fast&quot;, &quot;fast&quot;, &quot;slow&quot;, &quot;fast&quot;, &quot;fast&quot;]

      We will do the equivalent of:

          append(&apos;(&apos;);
          append(10);
          append(interleave(range, &apos; &apos;));
          append(10, &apos;)&apos;);

      This can be improved to find fast prefixes as well in
      the future.

* Tools/TestWebKitAPI/Tests/WTF/StringBuilder.cpp:
(TestWebKitAPI::TEST(StringBuilderTest, Interleave)):
    - Add tests for the three forms of interleave.

* Source/WTF/wtf/text/StringConcatenate.h:
(WTF::Interleave::writeUsing const):
(WTF::interleave):
    - Adds Interleave struct and introducing function.

(WTF::are8Bit):
    - Simplify to a single function using a fold.

(WTF::hasKnownLength):
(WTF::haveKnownLength):
    - Add helper to determine if the StringTypeAdapter has a known
      length to determine fast vs slow cases.

* Source/WebCore/css/CSSGradientValue.cpp:
(WebCore::CSSLinearGradientValue::customCSSText const):
(WebCore::CSSRadialGradientValue::customCSSText const):
(WebCore::CSSConicGradientValue::customCSSText const):
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::appendLangArgument):
(WebCore::appendLangArgumentList):
* Source/WebCore/css/CSSSelectorList.cpp:
(WebCore::CSSSelectorList::buildSelectorsText const):
* Source/WebCore/css/CSSSelectorList.h:
(WebCore::CSSSelectorList::const_iterator::operator!= const): Deleted.
* Source/WebCore/css/CSSValueList.cpp:
(WebCore::CSSValueContainingVector::serializeItems const):
* Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
(WebCore::CSSCalcOperationNode::buildCSSTextRecursive):
* Source/WebCore/css/query/MediaQueryParser.cpp:
(WebCore::MQ::serialize):
* Source/WebCore/css/typedom/transform/CSSTransformValue.cpp:
(WebCore::CSSTransformValue::serialize const):
* Source/WebCore/platform/text/TextFlags.cpp:
* Source/WebCore/testing/Internals.cpp:
(WebCore::serializeOffset):
(WebCore::serializeOffsets):
(WebCore::Internals::scrollSnapOffsets):
(WebCore::appendOffsets): Deleted.
    - Adopt interleave() for container serialization.

Canonical link: <a href="https://commits.webkit.org/279755@main">https://commits.webkit.org/279755@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e4d5217368b18be1156afbb5f652f16648f0f8f4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54410 "Failed to checkout and rebase branch from PR 29463") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33820 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6973 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57689 "Hash e4d52173 for PR 29463 does not build (failure)") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5141 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56712 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41355 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5104 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/57689 "Hash e4d52173 for PR 29463 does not build (failure)") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3449 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56504 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/31971 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47118 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/57689 "Hash e4d52173 for PR 29463 does not build (failure)") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/28777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4444 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3284 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/47771 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4659 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59280 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/53917 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/29630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/4765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51489 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/30788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47212 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50858 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11871 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/31765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/66216 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30580 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12617 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/text/StringConcatenate.h:
```diff
@@ -458,16 +458,140 @@ template<> class StringTypeAdapter<ASCIICaseConverter, void> {
 };
 
+template<typename C, typename E, typename B>
+struct Interleave {
```
Can we do this on one line?

darinadler commented on Source/WTF/wtf/text/StringConcatenate.h:
```diff
@@ -458,16 +458,140 @@ template<> class StringTypeAdapter<ASCIICaseConverter, void> {
+    const B& between;
+
+    template<typename Accumulator>
+    void writeUsing(Accumulator& accumulator) const
```
This too.

darinadler commented on Source/WTF/wtf/text/StringConcatenate.h:
```diff
@@ -458,16 +458,140 @@ template<> class StringTypeAdapter<ASCIICaseConverter, void> {
+//           ), ...
+//
+template<typename C, typename E, typename B>
+decltype(auto) interleave(C&& container, E&& each, B&& between)
```
Here too.

darinadler commented on Source/WTF/wtf/text/StringConcatenate.h:
```diff
@@ -458,16 +458,140 @@ template<> class StringTypeAdapter<ASCIICaseConverter, void> {
+}
+
+template<typename C, typename B>
+decltype(auto) interleave(C&& container, B&& between)
```
Here too.

darinadler commented on Source/WTF/wtf/text/StringConcatenate.h:
```diff
@@ -458,16 +458,140 @@ template<> class StringTypeAdapter<ASCIICaseConverter, void> {
+}
+
+template<typename C, typename E, typename B>
+class StringTypeAdapter<Interleave<C, E, B>, void> {
```
Here too.

darinadler commented on Source/WTF/wtf/text/StringConcatenate.h:
```diff
@@ -458,16 +458,140 @@ template<> class StringTypeAdapter<ASCIICaseConverter, void> {
+    }
+
+    template<typename Accumulator>
+    void writeUsing(Accumulator&& accumulator) const
```
Here too.

darinadler commented on Source/WTF/wtf/text/StringConcatenate.h:
```diff
@@ -458,16 +458,140 @@ template<> class StringTypeAdapter<ASCIICaseConverter, void> {
+};
+
+template<typename... Adapters>
+inline bool are8Bit(Adapters&& ...adapters)
```
Here too.

darinadler commented on Source/WTF/wtf/text/StringConcatenate.h:
```diff
@@ -458,16 +458,140 @@ template<> class StringTypeAdapter<ASCIICaseConverter, void> {
 {
-    return adapter.is8Bit();
+    return requires(Adapter x) {
+        { x.length() } -> std::same_as<unsigned>;
```
Could this subtly fail if `length()` returns `size_t` (or some other integer type) and cause us to take the slow path? And if so, how can we be sure to notice this if we make the error?

weinig replied:
Great point! Let me experiment and find out.

weinig replied:
Ok, I have two ideas:

1. Change std::same_as<unsigned> to `std::integral`, since I think that would cover all reasonable cases.

but I think it would be safe to error the other way and switch this to a check for the `writeUsing()` function that only the "slow" version has.

darinadler replied:
Either of those sounds good!

darinadler commented on Source/WebCore/css/CSSSelectorList.cpp:
```diff
@@ -115,12 +115,7 @@ String CSSSelectorList::selectorsText() const
-            stringBuilder.append(", "_s);
-        stringBuilder.append(subSelector->selectorText());
-    }
+    stringBuilder.append(interleave(*this, [](auto& subSelector) { return subSelector.selectorText(); }, ", "_s));
```
Sure wish this was `buildSelectorText(builder)` instead of `selectorText()`.

weinig replied:
Me too! I took a quick stab at it and it turned into a bit of an onion peel because of some weird-ish things selectorText() does right at the end:

```
        }
        return previousSelector->selectorText(separator, builder);
    } else if (auto separatorText = separatorTextForNestingRelative(); !separatorText.isNull()) {
        // We have a separator but no tag history which can happen with implicit relative nesting selector
        return separatorText + builder.toString();
    }

    return builder.toString();
```

Specifically that  "return separatorText + builder.toString();". I am sure there is a nice solution, but can be in its own PR.

darinadler commented on Source/WebCore/css/CSSSelectorList.h:
```diff
@@ -68,7 +68,7 @@ class CSSSelectorList {
         reference operator*() const { return *m_ptr; }
         pointer operator->() const { return m_ptr; }
-        bool operator!=(const const_iterator& other) const { return m_ptr != other.m_ptr; }
+        bool operator==(const const_iterator& other) const = default;
```
Omit the name `other` here?

darinadler commented on Source/WebCore/css/CSSValueList.cpp:
```diff
@@ -218,10 +218,7 @@ CSSValueListBuilder CSSValueContainingVector::copyValues() const
-    auto separator = separatorCSSText();
-    for (auto& value : *this)
-        builder.append(std::exchange(prefix, separator), value.cssText());
+    builder.append(interleave(*this, [](auto& value) { return value.cssText(); }, separatorCSSText()));
```
So irritating that this creates so many `String` temporaries.

weinig replied:
I’ve been pretty consistently adding both StringBuilder and non-StringBuilder versions for all the color value work, and we should propagate that idiom to all values.

darinadler commented on Source/WebCore/css/query/MediaQueryParser.cpp:
```diff
@@ -189,11 +189,9 @@ const FeatureSchema* MediaQueryParser::schemaForFeatureName(const AtomString& na
-            builder.append(", "_s);
-        serialize(builder, query);
-    }
+    // Trampoline is needed to avoid lookup ambiguity.
```
I’d like to understand more about that, and whether there are design changes we can make to make this less likely.

weinig replied:
Clang is unfortunately quite unhelpful with the error it gives, which is that it can't deduce "E" in interleave.

It comes down to the overload set containing multiple functions named `serialize` at the same level. I can also solve this by using a cast to the full function signature.

I think with some hand wringing we can probably construct overloads of interleave() that can do this, perhaps by having a version that looks something like:

```
template<>
decltype(auto) interleave(C&& container, (void(*)(StringBuilder&, const typename C::value_type&) each, B&& between) { ... }
```

or a variation on that using concepts, but I couldn't construct one that I was happy with.

I'll keep chipping away at it.

weinig replied:
I worked out a way to make this work for overloaded functions that take both a StringBuilder and value, but I still don't have a way to make it work if there are overloaded functions that take the value and return a string type.

The solution for the StringBuilder and value case was similar to what I put above. Overload interleave() with a second parameter of:

```c++
template<typename C> using EachTakingAccumulatorAndValueFunction = void(&)(StringBuilder&, const std::remove_reference_t<decltype(*std::begin(std::declval<C>()))>&);
 ```

looking like:

```c++
template<typename C, StringTypeAdaptable B> decltype(auto) interleave(const C& container, EachTakingAccumulatorAndValueFunction<C> each, B&& between)
{
    return Interleave {
        container,
        each,
        between
    };
}
```

The problem with the applying this same idea to the function returning a string value is that I don't know what the return type of that function would be and I can't figure out how to get c++ to deduce it for me.

darinadler reviewed (approved):
This is fantastic!

darinadler commented on Source/WTF/wtf/text/StringConcatenate.h:
```diff
@@ -44,6 +44,13 @@ WTF_EXPORT_PRIVATE extern std::atomic<int> wtfStringCopyCount;
 
+class StringBuilder;
+
+/// A type is `StringTypeAdaptable` if there is a specialization of `StringTypeAdapter` for that type.
```
What does the third slash mean?

Love the thoughtful use of concepts.

weinig replied:
Ah, this is what other projects use for more documentation style comments and headerdocs, and my mind cross-contaminating styles.

This should be // in WTF.

darinadler commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -5389,22 +5389,16 @@ MockContentFilterSettings& Internals::mockContentFilterSettings()
-        builder.append(coordinate.offset.toUnsigned());
-        if (coordinate.stop == ScrollSnapStop::Always)
-            builder.append(" (always)"_s);
+    builder.append(snapOffset.offset.toUnsigned());
+    if (snapOffset.stop == ScrollSnapStop::Always)
+        builder.append(" (always)"_s);
```
Could we do this with a single call to append?
```
builder.append(snapOffset.offset.toUnsigned(), snapOffset.stop == ScrollSnapStop::Always ? " (always)"_s : ""_s);
```

Commit:
Add support for easy serialization of container/ranges via StringBuilder
https://bugs.webkit.org/show_bug.cgi?id=275062

Reviewed by Darin Adler.

Introduced a new `StringTypeAdapter`, `Interleave`, which takes
a container/range, a joining string, and an optional mapping function.
`Interleaves` can currently only be used with StringBuilder, but can
be extended to makeString() as well with a bit of header factoring.

`Interleave` replaces common hand written loops when trying to
serialize container/ranges. For example, the following code:

    StringBuilder builder;

    ...

    builder.append('(');
    bool wroteFirstItem = false;
    for (auto& item : container) {
        if (wroteFirstItem)
            builder.append(", "_s);
        wroteFirstItem = true;
        builder.append(item->value);
    }
    builder.append(')');

can be replaced with:

    StringBuilder builder;

    ...

    builder.append('(', interleave(container, [](auto& item) { return item->value; }, ", "_s), ')');

`Interleave` works via a new kind of `StringTypeAdapter` that doesn't
know the length/8-bit-ness of its value before hand, causing the
builder to go down a new "slow" path. This "slow" path is just as
efficient as the old hand written loop, but a bit more ergonomic.

In addition to introducing `Interleave`, it has been adopted in places
in WebCore where it can be deployed (that I found), but has not been
adopted in places where adopting it would have turned the builder usage
into a one-liner. Those will be become makeString() calls when added.

* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::appendFromAdapters):
(WTF::StringBuilder::appendFromAdapterSlow):
(WTF::StringBuilder::appendFromAdaptersSlow):
    - Detect use of a "slow" `StringTypeAdapter` and use the
      alternative path, iteratively appending each adapter.

      The current strategy checks if any of the remaining
      `StringTypeAdapters` are "slow" and goes down the
      alternative path, peeling off the first adapter and
      appending it, and then recursing back to the beginning
      with the remainder of the adapters. This means that
      if an append() contains:

          append('(', 10, interleave(range, ' '), hex(10), ')'); [ aka "fast", "fast", "slow", "fast", "fast"]

      We will do the equivalent of:

          append('(');
          append(10);
          append(interleave(range, ' '));
          append(10, ')');

      This can be improved to find fast prefixes as well in
      the future.

* Tools/TestWebKitAPI/Tests/WTF/StringBuilder.cpp:
(TestWebKitAPI::TEST(StringBuilderTest, Interleave)):
    - Add tests for the three forms of interleave.

* Source/WTF/wtf/text/StringConcatenate.h:
(WTF::Interleave::writeUsing const):
(WTF::interleave):
    - Adds Interleave struct and introducing function.

(WTF::are8Bit):
    - Simplify to a single function using a fold.

(WTF::hasKnownLength):
(WTF::haveKnownLength):
    - Add helper to determine if the StringTypeAdapter has a known
      length to determine fast vs slow cases.

* Source/WebCore/css/CSSGradientValue.cpp:
(WebCore::CSSLinearGradientValue::customCSSText const):
(WebCore::CSSRadialGradientValue::customCSSText const):
(WebCore::CSSConicGradientValue::customCSSText const):
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::appendLangArgument):
(WebCore::appendLangArgumentList):
* Source/WebCore/css/CSSSelectorList.cpp:
(WebCore::CSSSelectorList::buildSelectorsText const):
* Source/WebCore/css/CSSSelectorList.h:
(WebCore::CSSSelectorList::const_iterator::operator!= const): Deleted.
* Source/WebCore/css/CSSValueList.cpp:
(WebCore::CSSValueContainingVector::serializeItems const):
* Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
(WebCore::CSSCalcOperationNode::buildCSSTextRecursive):
* Source/WebCore/css/query/MediaQueryParser.cpp:
(WebCore::MQ::serialize):
* Source/WebCore/css/typedom/transform/CSSTransformValue.cpp:
(WebCore::CSSTransformValue::serialize const):
* Source/WebCore/platform/text/TextFlags.cpp:
* Source/WebCore/testing/Internals.cpp:
(WebCore::serializeOffset):
(WebCore::serializeOffsets):
(WebCore::Internals::scrollSnapOffsets):
(WebCore::appendOffsets): Deleted.
    - Adopt interleave() for container serialization.

Canonical link: https://commits.webkit.org/279755@main
